[DOCU-2601] Secrets management with decK (#4661)
Kong/docs.konghq.com: src/deck/guides/vaults.md

You can store your secrets in a [vault backend](/gateway/latest/kong-enterprise/secrets-management/),
then reference them in your declarative configuration files.

You can use secrets to store sensitive data, such as credentials.

See [Secrets Management in {{site.base_gateway}}](/gateway/latest/kong-enterprise/secrets-management/#what-can-be-stored-as-a-secret)
for a full list of values that can be stored as secrets.

{:.note}
> For storing configuration values as environment variables on the node running decK,
see [Using Environment Variables with decK](/deck/latest/guides/environment-variables/).
The reference format for secrets is _not_ the same as references for environment
variables used by decK.

## Configure a secret vault

Set up a secret vault using the {{site.base_gateway}} `vaults` entity.

For example, add the following snippet to your
declarative configuration file (`kong.yaml` by default) to set up a vault using
environment variables as the backend, configure a prefix for the vault, and a
prefix for the reference:

```yaml
_format_version: "3.0"
vaults:
- config:
    prefix: MY_SECRET_
  description: ENV vault for secrets
  name: env
  prefix: my-env-vault
```

> Key | Description

Reviewer: I think you should wrap the keys in back ticks

----|---

> vauts.config | Stores the configuration for a particular vault. The configuration values required depend on the vault that you are using. In this example, the `vaults.config.prefix` value configures the prefix for the environment variable that the value will be stored in. See the individual [vault backends](/gateway/latest/kong-enterprise/secrets-management/backends/) to find the required configuration values for your particular vault type.

Reviewer: ```suggestion
vaults.config | Stores the configuration for a particular vault. The configuration values required depend on the vault that you are using. In this example, the `vaults.config.prefix` value configures the prefix for the environment variable that the value will be stored in. See the individual [vault backends](/gateway/latest/kong-enterprise/secrets-management/backends/) to find the required configuration values for your particular vault type.
```

---
title: Secret Management with decK
content_type: how-to
---

decK supports secret references for encoded values in {{site.base_gateway}}.
You can store your secrets a [vault backend](/gateway/latest/kong-enterprise/secrets-management/),
then reference them in your declarative configuration files.

This guide covers storing values as secrets. You can use secrets to store sensitive data, such as credentials.

## {{site.base_gateway}} secrets or decK environment variables?

With decK, instead of checking credentials into Git, you can choose one of the following options:

Option | Description | Why use this method?
-------|-------------|---------------------
[decK environment variables](/deck/latest/guides/environment-variables) | Store values as environment variables and access them directly through decK. {{site.base_gateway}} is not involved. | • You can use this option for environment-specific values. <br><br> • This method can store any configuration values used by {{site.base_gateway}} entities. It can't secure values from `kong.conf`. <br><br> • Available for all {{site.base_gateway}} packages: open-source, Enterprise Free mode, and Enterprise licensed mode.

> Secrets in {{site.base_gateway}} | Store values as secrets in a vault, then reference the secrets with a vault reference. In this case, {{site.base_gateway}} manages the secrets with a `vaults` entity. | • This option is more secure. <br><br> • You can use secrets to store many sensitive values, including parameters in Kong's configuration (`kong.conf`). See [Secrets Management in {{site.base_gateway}}](/gateway/latest/kong-enterprise/secrets-management/#what-can-be-stored-as-a-secret) for a full list. <br><br> • Secrets management is only available for {{site.base_gateway}} Enterprise packages. It is not available for open-source {{site.base_gateway}}. <br>The environment variable vault can be used in Free mode without a license, while all other vault backends require a license.

Reviewer: ```suggestion
Secrets in {{site.base_gateway}} | Store values as secrets in a vault, then reference the secrets with a vault backend. In this case, the {{site.base_gateway}} data plane manages the secrets with a `vaults` entity. |  • You can use secrets to store sensitive values, see [Secrets Management in {{site.base_gateway}}](/gateway/latest/kong-enterprise/secrets-management/#what-can-be-stored-as-a-secret) for a full list. <br><br> • Secrets management is only available for {{site.base_gateway}} Enterprise packages. It is not available for open-source {{site.base_gateway}}. <br>The environment variable vault can be used in Free mode without a license, while all other vault backends require a license.
```

Reviewer: I think the license stuff should move to the description box. I think the why could stand to list the supported backends, site security independently, like "Is a secure way to manage sensitive information".

configure a prefix for the vault, and a prefix for the reference:

```yaml
_format_version: "3.0"
vaults:
- config:
    prefix: MY_SECRET_
  description: ENV vault for secrets
  name: env
  prefix: my-env-vault
```

Key | Description
----|---
vauts.config | Stores the configuration for a particular vault. The configuration values required depend on the vault that you are using. In this example, the `vaults.config.prefix` value configures the prefix for the environment variable that the value will be stored in. See the individual [vault backends](/gateway/latest/kong-enterprise/secrets-management/backends/) to find the required configuration values for your particular vault type.
vaults.description | An optional description for your vault.
vaults.name | The type of vault. Accepts one of: `env`, `gcp`, `aws`, or `hcv`.
vaults.prefix | The reference prefix. You need this prefix to access secrets stored in this vault. For example, `{vault://my-env-vault/<some-secret>}`.

{{site.base_gateway}} also supports HashiCorp Vault, GCP, and AWS as [vault backends](/gateway/latest/kong-enterprise/secrets-management/backends/).

{:.note}

> > **Note**: We recommend splitting out `vault` declarative configuration using [distributed configuration](/deck/latest/guides/distributed-configuration).

Reviewer 2: Instead of adding this as a note, can we expand it into a best-practice/recommended section? I believe that's super important to make users understand the implications and risks how mixing declarative configuration of vaults with other regular Kong entities and how to prevent their misusage.

Author: > understand the implications and risks how mixing declarative configuration of vaults with other regular Kong entities and how to prevent their misusage.

Can you point me to somewhere that I can track down this information?

Reviewer 3: > Can you elaborate on why it's bad to mix these configurations?

1. Vault are closer to "infrastructure" than other Kong configurations. Separation of routing policies from infrastructure-specific bit is going to be helpful.
2. Vaults may be shared across teams. In this case, one specific team shouldn't control its configuration. One team changing the vault a can have disastrous impact on another team.
3. Vaults if deleted while they are in use (use meaning references to secrets in configuration via the vault exist) can lead to total loss of proxy capabilities.

If you are a small team that operators and configures Kong, then mixing them is fine.

> Expanding on the above, should users split only the vaults entity config, or should each entity using a vault reference also be tagged?

Only the vault.

> When you say split out vaults config, do you mean with tags only?

1. the "distributed configuration via select tags" feature
5. recommend a separate RBAC role, user, and token
6. recommend a separate CI pipeline for vaults

## Store and reference secrets

> Store your sensitive values in a secret:

Reviewer 2: I think it's important to note that these env variables should reside in the Gateway instance and not in the node executing decK.

> Store your sensitive values in a secret:

Reviewer: ```suggestion
Store your sensitive values as a secret:
```
